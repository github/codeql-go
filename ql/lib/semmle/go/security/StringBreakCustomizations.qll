/**
 * Provides default sources, sinks and sanitizers for reasoning about unsafe-quoting
 * vulnerabilities, as well as extension points for adding your own.
 */

import go

/**
 * Provides extension points for customizing the taint tracking configuration for reasoning about
 * unsafe-quoting vulnerabilities.
 */
module StringBreak {
  /** A (single or double) quote. */
  class Quote extends string {
    Quote() { this = "'" or this = "\"" }

    /** Gets the type of this quote, either single or double. */
    string getType() { if this = "'" then result = "single" else result = "double" }
  }

  /**
   * A data flow source for unsafe-quoting vulnerabilities.
   */
  abstract class Source extends DataFlow::Node { }

  /**
   * A data flow sink for unsafe-quoting vulnerabilities.
   */
  abstract class Sink extends DataFlow::Node {
    /** Gets the quote character for which this is a sink. */
    abstract Quote getQuote();
  }

  /**
   * A sanitizer for unsafe-quoting vulnerabilities.
   */
  abstract class Sanitizer extends DataFlow::Node {
    /** Gets the quote character for which this is a sanitizer. */
    Quote getQuote() { any() }
  }

  /**
   * A sanitizer guard for unsafe-quoting vulnerabilities.
   */
  abstract class SanitizerGuard extends DataFlow::BarrierGuard { }

  /** Holds if `l` contains a `quote` (either single or double). */
  private predicate containsQuote(StringOps::ConcatenationLeaf l, Quote quote) {
    quote = l.getStringValue().regexpFind("['\"]", _, _)
  }

  /** A call to `json.Marshal`, considered as a taint source for unsafe quoting. */
  class JsonMarshalAsSource extends Source {
    JsonMarshalAsSource() {
      exists(MarshalingFunction jsonMarshal | jsonMarshal.getFormat() = "JSON" |
        this = jsonMarshal.getOutput().getNode(jsonMarshal.getACall())
      )
    }
  }

  /** A string concatenation with quotes, considered as a taint sink for unsafe quoting. */
  class StringConcatenationAsSink extends Sink {
    Quote quote;

    StringConcatenationAsSink() {
      exists(StringOps::ConcatenationLeaf lf | lf.asNode() = this |
        containsQuote(lf.getPreviousLeaf(), quote) and
        containsQuote(lf.getNextLeaf(), quote)
      )
    }

    override Quote getQuote() { result = quote }
  }

  /** A call to `json.Unmarshal`, considered as a sanitizer for unsafe quoting. */
  class UnmarshalSanitizer extends Sanitizer {
    UnmarshalSanitizer() {
      exists(UnmarshalingFunction jsonUnmarshal | jsonUnmarshal.getFormat() = "JSON" |
        this = jsonUnmarshal.getOutput().getNode(jsonUnmarshal.getACall())
      )
    }
  }

  /**
   * A call to `strings.Replace` or `strings.ReplaceAll`, considered as a sanitizer
   * for unsafe quoting.
   */
  class ReplaceSanitizer extends Sanitizer {
    Quote quote;

    ReplaceSanitizer() {
      exists(string name, DataFlow::CallNode call |
        this = call and
        call.getTarget().hasQualifiedName("strings", name) and
        call.getArgument(1).getStringValue().matches("%" + quote + "%")
      |
        name = "Replace" and
        call.getArgument(3).getNumericValue() < 0
        or
        name = "ReplaceAll"
      )
    }

    override Quote getQuote() { result = quote }
  }

  class StringsNewReplacerCall extends DataFlow::CallNode {
    StringsNewReplacerCall() { this.getTarget().hasQualifiedName("strings", "NewReplacer") }

    DataFlow::Node getAReplacedArgument() {
      exists(int m, int n | m = 2 * n and n = m / 2 and result = getArgument(m))
    }
  }

  class StringsNewReplacerConfiguration extends DataFlow2::Configuration {
    StringsNewReplacerConfiguration() { this = "StringsNewReplacerConfiguration" }

    override predicate isSource(DataFlow::Node source) { source instanceof StringsNewReplacerCall }

    override predicate isSink(DataFlow::Node sink) {
      exists(DataFlow::MethodCallNode call |
        sink = call.getReceiver() and
        call.getTarget().hasQualifiedName("strings", "Replacer", ["Replace", "WriteString"])
      )
    }
  }

  /**
   * A call to `strings.Replacer.Replace`, considered as a sanitizer for unsafe
   * quoting.
   */
  class ReplacerReplaceSanitizer extends DataFlow::MethodCallNode, Sanitizer {
    Quote quote;

    ReplacerReplaceSanitizer() {
      exists(StringsNewReplacerConfiguration config, DataFlow::Node source, DataFlow::Node sink |
        config.hasFlow(source, sink) and
        this.getTarget().hasQualifiedName("strings", "Replacer", "Replace") and
        sink = this.getReceiver() and
        quote = source.(StringsNewReplacerCall).getAReplacedArgument().getStringValue()
      )
    }

    override Quote getQuote() { result = quote }
  }

  /**
   * A call to `strings.Replacer.WriteString`, considered as a sanitizer for
   * unsafe quoting.
   */
  class ReplacerWriteStringSanitizer extends Sanitizer {
    Quote quote;

    ReplacerWriteStringSanitizer() {
      exists(
        StringsNewReplacerConfiguration config, DataFlow::Node source, DataFlow::Node sink,
        DataFlow::MethodCallNode call
      |
        config.hasFlow(source, sink) and
        call.getTarget().hasQualifiedName("strings", "Replacer", "WriteString") and
        sink = call.getReceiver() and
        this = call.getArgument(1) and
        quote = source.(StringsNewReplacerCall).getAReplacedArgument().getStringValue()
      )
    }

    override Quote getQuote() { result = quote }
  }
}
