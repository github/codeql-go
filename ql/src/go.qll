/**
 * Provides classes for working with Go programs.
 */

import Customizations
import semmle.go.AST
import semmle.go.Comments
import semmle.go.Concepts
import semmle.go.Decls
import semmle.go.Expr
import semmle.go.Files
import semmle.go.GoModExpr
import semmle.go.Locations
import semmle.go.Packages
import semmle.go.Scopes
import semmle.go.Stmt
import semmle.go.StringOps
import semmle.go.Types
import semmle.go.controlflow.BasicBlocks
import semmle.go.controlflow.ControlFlowGraph
import semmle.go.controlflow.IR
import semmle.go.dataflow.DataFlow
import semmle.go.dataflow.GlobalValueNumbering
import semmle.go.dataflow.TaintTracking
import semmle.go.dataflow.SSA
import semmle.go.frameworks.HTTP
import semmle.go.frameworks.SystemCommandExecutors
import semmle.go.frameworks.SQL
import semmle.go.frameworks.Stdlib
import semmle.go.security.FlowSources
import semmle.go.Util
