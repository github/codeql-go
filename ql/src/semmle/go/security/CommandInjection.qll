/**
 * Provides a taint tracking configuration for reasoning about command
 * injection vulnerabilities.
 *
 * Note, for performance reasons: only import this file if
 * `CommandInjection::Configuration` is needed, otherwise
 * `CommandInjectionCustomizations` should be imported instead.
 */

import go

/**
 * Provides a taint tracking configuration for reasoning about command
 * injection vulnerabilities.
 */
module CommandInjection {
  import CommandInjectionCustomizations::CommandInjection

  /**
   * A taint-tracking configuration for reasoning about command-injection vulnerabilities.
   */
  class Configuration extends TaintTracking::Configuration {
    Configuration() { this = "CommandInjection" }

    override predicate isSource(DataFlow::Node source) { source instanceof Source }

    override predicate isSink(DataFlow::Node sink) {
      exists(Sink s | sink = s | not s.doubleDashIsSanitizing())
    }

    override predicate isSanitizer(DataFlow::Node node) {
      super.isSanitizer(node) or
      node instanceof Sanitizer
    }

    override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
      guard instanceof SanitizerGuard
    }
  }

  private class ArgumentArrayWithDoubleDash extends DataFlow::Node {
    int doubleDashIndex;

    ArgumentArrayWithDoubleDash() {
      // Call whose argument list contains a "--"
      exists(DataFlow::CallNode c |
        this = c and
        (c = Builtin::append().getACall() or c = any(SystemCommandExecution sce)) and
        c.getArgument(doubleDashIndex).getStringValue() = "--"
      )
      or
      // array/slice literal containing a "--"
      exists(ArrayOrSliceLit litExpr |
        this = DataFlow::exprNode(litExpr) and
        litExpr.getElement(doubleDashIndex).getStringValue() = "--"
      )
      or
      // call consuming an existing an array with a "--"
      exists(ArgumentArrayWithDoubleDash alreadyHasDoubleDash, DataFlow::CallNode userCall |
        (
          alreadyHasDoubleDash.getType().getUnderlyingType() instanceof ArrayType or
          alreadyHasDoubleDash.getType() instanceof SliceType
        ) and
        this = userCall and
        DataFlow::localFlow(alreadyHasDoubleDash, userCall.getArgument(doubleDashIndex))
      )
    }

    DataFlow::Node getASanitizedElement() {
      exists(int sanitizedIndex |
        sanitizedIndex > doubleDashIndex and
        (
          result = this.(DataFlow::CallNode).getArgument(sanitizedIndex) or
          result = DataFlow::exprNode(this.asExpr().(ArrayOrSliceLit).getElement(sanitizedIndex))
        )
      )
    }
  }

  class DoubleDashSanitizingConfiguration extends TaintTracking::Configuration {
    DoubleDashSanitizingConfiguration() { this = "CommandInjectionWithDoubleDashSanitizer" }

    override predicate isSource(DataFlow::Node source) { source instanceof Source }

    override predicate isSink(DataFlow::Node sink) {
      exists(Sink s | sink = s | s.doubleDashIsSanitizing())
    }

    override predicate isSanitizer(DataFlow::Node node) {
      super.isSanitizer(node) or
      node instanceof Sanitizer or
      node = any(ArgumentArrayWithDoubleDash array).getASanitizedElement()
    }

    override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
      guard instanceof SanitizerGuard
    }

    // Hack: with use-use flow, we might have x (use at line 1) -> x (use at line 2),
    // x (use at line 1) -> array at line 1 and x (use at line 2) -> array at line 2,
    // in the context
    //
    // array1 := {"--", x}
    // array2 := {x, "--"}
    //
    // We want to taint array2 but not array1, which suggests excluding the edge x (use 1) -> array1
    // However isSanitizer only allows us to remove nodes (isSanitizerIn/Out permit removing all outgoing
    // or incoming edges); we can't remove an individual edge, so instead we supply extra edges connecting
    // the definition with the next use.
    override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
      exists(
        ArgumentArrayWithDoubleDash array, DataFlow::InstructionNode sanitized,
        DataFlow::SsaNode defn
      |
        sanitized = array.getASanitizedElement() and sanitized = defn.getAUse()
      |
        pred = defn and succ = sanitized.getASuccessor()
      )
    }
  }
}
