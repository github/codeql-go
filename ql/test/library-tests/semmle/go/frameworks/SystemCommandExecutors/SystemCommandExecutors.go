package main

import (
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path"
	"strings"

	"github.com/codeskyblue/go-sh"
	"golang.org/x/crypto/ssh"
)

func main() {}

func handler(w http.ResponseWriter, req *http.Request) {
	sudo := "sudo"
	shell := "/bin/bash"
	assumedNonShell := "ls"
	args := []string{}

	source := req.URL.Query()["cmd"][0]

	// os.StartProcess: these MUST be caught.
	{
		// `source` is used directly as command:
		os.StartProcess(source, args, nil)

		// `source` flows into a composite literal which is used
		// as arguments, and the command is a shell:
		os.StartProcess(shell, []string{source}, nil)

		// `source` flows into a composite literal as first argument to append:
		os.StartProcess(shell, append([]string{source}, args...), nil)

		// `source` flows into a composite literal as Nth argument to append:
		os.StartProcess(shell, append([]string{sudo}, source), nil)

		// TODO:
		//shellToAssumedNonShell := append([]string{sudo}, shell, source)
		//os.StartProcess(assumedNonShell, shellToAssumedNonShell, nil)
	}

	// os.StartProcess: these MUST NOT be caught because non-valid.
	{
		// `source` is an argument to a non-shell command that does not execute
		// the `source` as a command, i.e. the source is just an argument to a command
		// that will not execute it.
		os.StartProcess(assumedNonShell, []string{source}, nil)

		// as above, except the source is inside a composite literal inside an append:
		os.StartProcess(assumedNonShell, append([]string{source}, args...), nil)

		// source is used inside append as nth argument:
		os.StartProcess(assumedNonShell, append([]string{assumedNonShell}, source), nil)
	}

	// exec.Command: these MUST be caught.
	{
		// source is used directly as command:
		exec.Command(source, args...).Run()

		// source comes as nth arg to a shell:
		exec.Command(shell, "a0", "a1", source)

		// source flows into a composite literal used as args, and the command is a shell:
		exec.Command(shell, []string{"a0", "a1", source}...)

		// source flows into a composite literal as first argument to append:
		exec.Command(shell, append([]string{source}, args...)...)

		// source flows into a composite literal as Nth argument to append:
		exec.Command(shell, append([]string{sudo}, source)...)

		// other ways to compose a command:
		exec.Command("sh", "-c", source)
		exec.Command("sudo", "sh", "-c", source)

		// (this is a shell, so you can just create a pipe with `|` or a `$(command)` or a `&& command`):
		{
			// string concatenation:
			exec.Command("sh", "-c", "GOOS=windows GOARCH=386 go build -ldflags \"-s -w -H=windowsgui\" -o \""+source+".go")

			// source is argument to fmt.Sprintf:
			exec.Command("/bin/bash", "-c", fmt.Sprintf("ls %s", source))

			// source passes through append and strings.Join:
			exec.Command("sh", "-c", strings.Join(append([]string{assumedNonShell}, source), " "))
			exec.Command("sh", "-c", strings.Join(append([]string{assumedNonShell}, "a0", "a1", source), " "))

			// source passes through composite literal, append and strings.Join:
			exec.Command("sh", "-c", assumedNonShell, strings.Join(append([]string{source}, "a0", "a1"), " "))

			// source passes through path.Join:
			exec.Command("/bin/sh", path.Join("scripts", source))
		}
	}
	// golang.org/x/crypto/ssh
	{
		session := &ssh.Session{}
		session.CombinedOutput(source)
		session.Output(source)
		session.Run(source)
		session.Start(source)
	}
	// github.com/codeskyblue/go-sh
	{
		sh.Command(shell, toInterfaceArray(append([]string{assumedNonShell}, source)...)...)
		sh.InteractiveSession().Call(shell, toInterfaceArray(append([]string{assumedNonShell}, source)...)...)
		sh.InteractiveSession().Command(shell, toInterfaceArray(append([]string{assumedNonShell}, source)...)...)
	}
}
func toInterfaceArray(str ...string) []interface{} {
	res := make([]interface{}, 0)
	for i := range str {
		res = append(res, str[i])
	}
	return res
}
