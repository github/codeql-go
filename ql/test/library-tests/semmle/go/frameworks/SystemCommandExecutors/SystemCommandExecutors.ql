/**
 * @name Command built from user-controlled sources
 * @description Building a system command from user-controlled sources is vulnerable to insertion of
 *              malicious code by the user.
 * @kind path-problem
 * @problem.severity error
 * @id go/command-injection
 * @tags security
 *       external/cwe/cwe-078
 */

import go
import semmle.go.security.CommandInjection
import DataFlow::PathGraph

// "net/http".Request sources:
class NethttpRequestSource extends UntrustedFlowSource::Range, DataFlow::Node {
  string thisPackagePath;
  string thisTypeName;

  NethttpRequestSource() {
    thisPackagePath = "net/http" and
    thisTypeName = "Request" and
    (
      // method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(thisPackagePath, thisTypeName, methodName) and
        (
          methodName = "BasicAuth" and
          (
            //### 0 string string
            resultIndex = 0
            or
            //### 1 string string
            resultIndex = 1
          )
          or
          methodName = "FormValue" and
          //### 0 string string
          resultIndex = 0
          or
          methodName = "UserAgent" and
          //### 0 string string
          resultIndex = 0
          or
          methodName = "Referer" and
          //### 0 string string
          resultIndex = 0
          or
          methodName = "PostFormValue" and
          //### 0 string string
          resultIndex = 0
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
      or
      // field reads:
      exists(DataFlow::FieldReadNode fieldRead, DataFlow::Field fld, string fieldName |
        fieldRead = fld.getARead() and
        (
          fld.hasQualifiedName(thisPackagePath, thisTypeName, fieldName) and
          (
            //### string string
            fieldName = "Host"
            or
            //### string string
            fieldName = "Method"
            or
            //### string string
            fieldName = "Proto"
            or
            //### string string
            fieldName = "RequestURI"
            or
            //### string string
            fieldName = "RemoteAddr"
            or
            //### slice []string
            fieldName = "TransferEncoding"
          )
        )
      |
        this = fieldRead
      )
    )
  }
}

// "net/http".Cookie sources:
class NethttpCookieSource extends UntrustedFlowSource::Range, DataFlow::Node {
  string thisPackagePath;
  string thisTypeName;

  NethttpCookieSource() {
    thisPackagePath = "net/http" and
    thisTypeName = "Cookie" and
    (
      // method calls:
      exists(DataFlow::MethodCallNode call, string methodName, int resultIndex |
        call.getTarget().hasQualifiedName(thisPackagePath, thisTypeName, methodName) and
        (
          methodName = "String" and
          //### 0 string string
          resultIndex = 0
        )
      |
        resultIndex = 0 and this = call.getResult()
        or
        this = call.getResult(resultIndex)
      )
      or
      // field reads:
      exists(DataFlow::FieldReadNode fieldRead, DataFlow::Field fld, string fieldName |
        fieldRead = fld.getARead() and
        (
          fld.hasQualifiedName(thisPackagePath, thisTypeName, fieldName) and
          (
            //### slice []string
            fieldName = "Unparsed"
            or
            //### string string
            fieldName = "Value"
            or
            //### string string
            fieldName = "Path"
            or
            //### string string
            fieldName = "Domain"
            or
            //### string string
            fieldName = "RawExpires"
            or
            //### string string
            fieldName = "Raw"
            or
            //### string string
            fieldName = "Name"
          )
        )
      |
        this = fieldRead
      )
    )
  }
}

/**
 * ShellOrSudoExecution extends system command execution sinks to include
 * arguments passed to command interpreter binaries like shells, sudo,
 * or programming language interpreters.
 */
class ShellOrSudoExecution extends SystemCommandExecution::Range, DataFlow::CallNode {
  DataFlow::ExprNode argumentNode;

  ShellOrSudoExecution() {
    exists(SystemCommandExecution exec |
      // Either there is another argument to this call that is a sudo/shell ...
      exists(ShellLike shellOrSudoArg |
        shellOrSudoArg = exec.(DataFlow::CallNode).getAnArgument().getAPredecessor*()
      |
        argumentNode = exec.(DataFlow::CallNode).getAnArgument()
      )
    |
      this = exec
    )
  }

  override DataFlow::Node getCommandName() { result = argumentNode }
}

class SystemCommandExecutors extends SystemCommandExecution::Range, DataFlow::CallNode {
  int cmdArg;

  SystemCommandExecutors() {
    exists(string pkg, string name | this.getTarget().hasQualifiedName(pkg, name) |
      pkg = "os" and name = "StartProcess" and cmdArg = 0
      or
      // assume that if a `Cmd` is instantiated it will be run
      pkg = "os/exec" and name = "Command" and cmdArg = 0
      or
      pkg = "os/exec" and name = "CommandContext" and cmdArg = 1
    )
  }

  override DataFlow::Node getCommandName() { result = this.getArgument(cmdArg) }
}

/**
 * A call to the `Command` function, or `Call` or `Command` methods on a `Session` object
 * from the [go-sh](https://github.com/codeskyblue/go-sh) package, viewed as a
 * system-command execution.
 */
class GoShCommandExecution extends SystemCommandExecution::Range, DataFlow::CallNode {
  GoShCommandExecution() {
    exists(string packagePath |
      packagePath = "github.com/codeskyblue/go-sh" and
      (
        // Catch method calls on the `Session` object:
        exists(Method method |
          method.hasQualifiedName(packagePath, "Session", "Call")
          or
          method.hasQualifiedName(packagePath, "Session", "Command")
        |
          this = method.getACall()
        )
        or
        // Catch calls to the `Command` function:
        getTarget().hasQualifiedName(packagePath, "Command")
      )
    )
  }

  override DataFlow::Node getCommandName() { result = this.getArgument(0) }
}

/**
 * A call to a method on a `Session` object from the [ssh](golang.org/x/crypto/ssh)
 * package, viewed as a system-command execution.
 */
class SshCommandExecution extends SystemCommandExecution::Range, DataFlow::CallNode {
  SshCommandExecution() {
    // Catch method calls on the `Session` object:
    exists(Method method, string methodName |
      method.hasQualifiedName("golang.org/x/crypto/ssh", "Session", methodName) and
      methodName = "CombinedOutput"
      or
      methodName = "Output"
      or
      methodName = "Run"
      or
      methodName = "Start"
    |
      this = method.getACall()
    )
  }

  override DataFlow::Node getCommandName() { result = this.getArgument(0) }
}

/**
 * ShellLike defines a string value node whose value is any binary
 * (shells, sudo, programming language interpreters, ssh clients, etc.)
 * which executes arguments passed to as commands.
 */
class ShellLike extends DataFlow::Node {
  ShellLike() {
    isSudoOrSimilar(this) or
    isShell(this) or
    isProgrammingLanguageCli(this) or
    isSsh(this)
  }
}

string getASudoCommand() {
  result = "sudo" or
  result = "su" or
  result = "doas" or
  result = "access" or
  result = "vsys" or
  result = "userv" or
  result = "sus" or
  result = "super" or
  result = "priv" or
  result = "calife" or
  result = "ssu" or
  result = "su1" or
  result = "op" or
  result = "sudowin" or
  result = "sudown"
}

/**
 * isSudoOrSimilar defines a string value node whose value is any binary
 * that work in a similar manner to `sudo`, whose arguments are
 * interpreted as system commands.
 */
predicate isSudoOrSimilar(DataFlow::Node node) {
  exists(string regex |
    regex = ".*(^|/)(" + concat(string cmd | cmd = getASudoCommand() | cmd, "|") + ")"
  |
    node.getStringValue().regexpMatch(regex)
  )
}

string getAShellCommand() {
  result = "bash" or
  result = "sh" or
  result = "rbash" or
  result = "dash" or
  result = "zsh" or
  result = "csh" or
  result = "tcsh" or
  result = "fish" or
  result = "pwsh" or
  result = "elvish" or
  result = "oh" or
  result = "ion" or
  result = "ksh" or
  result = "rksh" or
  result = "tksh" or
  result = "mksh" or
  result = "nu" or
  result = "oksh" or
  result = "osh" or
  result = "shpp" or
  result = "xiki" or
  result = "xonsh" or
  result = "yash"
}

/**
 * isShell defines a string value node whose value is any shell,
 * whose arguments are interpreted as system commands.
 */
predicate isShell(DataFlow::Node node) {
  exists(string regex |
    regex = ".*(^|/)(" + concat(string cmd | cmd = getAShellCommand() | cmd, "|") + ")"
  |
    node.getStringValue().regexpMatch(regex)
  )
}

string getAnInterpreterName() {
  result = "python" or
  result = "php" or
  result = "ruby" or
  result = "perl" or
  result = "node" or
  result = "nodejs"
}

/**
 * isShell defines a string value node whose value is any CLI
 * programming language interpreter, whose arguments are
 * executed as code.
 */
predicate isProgrammingLanguageCli(DataFlow::Node node) {
  // NOTE: we can enounter cases like /usr/bin/python3.1
  node.getStringValue().matches("%/python%")
  or
  exists(string regex |
    regex =
      ".*(^|/)(" + concat(string cmd | cmd = getAnInterpreterName() | cmd + "[\\d.\\-v]*", "|") + ")"
  |
    node.getStringValue().regexpMatch(regex)
  )
}

string getASshCommand() { result = "ssh" or result = "putty.exe" or result = "kitty.exe" }

/**
 * isSsh defines a string value node whose value is a ssh client or similar,
 * whose arguments can be commands that will be executed on the remote host.
 */
predicate isSsh(DataFlow::Node node) {
  exists(string regex |
    regex = ".*(^|/)(" + concat(string cmd | cmd = getASshCommand() | cmd, "|") + ")"
  |
    node.getStringValue().regexpMatch(regex)
  )
}

from CommandInjection::Configuration cfg, DataFlow::PathNode source, DataFlow::PathNode sink
where cfg.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "This command depends on $@.", source.getNode(),
  "a user-provided value"
