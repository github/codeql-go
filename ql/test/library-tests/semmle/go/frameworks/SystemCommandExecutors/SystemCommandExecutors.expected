edges
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:24:19:24:24 | source |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:28:26:28:41 | composite literal |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:31:26:31:58 | call to append |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:34:26:34:55 | call to append |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:58:16:58:21 | source |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:61:35:61:40 | source |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:64:23:64:50 | composite literal |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:67:23:67:55 | call to append |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:70:23:70:52 | call to append |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:73:28:73:33 | source |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:74:36:74:41 | source |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:79:29:79:114 | ...+... |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:82:36:82:63 | call to Sprintf |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:85:29:85:88 | call to Join |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:86:29:86:100 | call to Join |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:89:46:89:100 | call to Join |
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:92:28:92:55 | call to Join |
nodes
| SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | semmle.label | selection of URL : pointer type |
| SystemCommandExecutors.go:24:19:24:24 | source | semmle.label | source |
| SystemCommandExecutors.go:28:26:28:41 | composite literal | semmle.label | composite literal |
| SystemCommandExecutors.go:31:26:31:58 | call to append | semmle.label | call to append |
| SystemCommandExecutors.go:34:26:34:55 | call to append | semmle.label | call to append |
| SystemCommandExecutors.go:58:16:58:21 | source | semmle.label | source |
| SystemCommandExecutors.go:61:35:61:40 | source | semmle.label | source |
| SystemCommandExecutors.go:64:23:64:50 | composite literal | semmle.label | composite literal |
| SystemCommandExecutors.go:67:23:67:55 | call to append | semmle.label | call to append |
| SystemCommandExecutors.go:70:23:70:52 | call to append | semmle.label | call to append |
| SystemCommandExecutors.go:73:28:73:33 | source | semmle.label | source |
| SystemCommandExecutors.go:74:36:74:41 | source | semmle.label | source |
| SystemCommandExecutors.go:79:29:79:114 | ...+... | semmle.label | ...+... |
| SystemCommandExecutors.go:82:36:82:63 | call to Sprintf | semmle.label | call to Sprintf |
| SystemCommandExecutors.go:85:29:85:88 | call to Join | semmle.label | call to Join |
| SystemCommandExecutors.go:86:29:86:100 | call to Join | semmle.label | call to Join |
| SystemCommandExecutors.go:89:46:89:100 | call to Join | semmle.label | call to Join |
| SystemCommandExecutors.go:92:28:92:55 | call to Join | semmle.label | call to Join |
#select
| SystemCommandExecutors.go:24:19:24:24 | source | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:24:19:24:24 | source | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:28:26:28:41 | composite literal | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:28:26:28:41 | composite literal | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:31:26:31:58 | call to append | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:31:26:31:58 | call to append | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:34:26:34:55 | call to append | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:34:26:34:55 | call to append | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:58:16:58:21 | source | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:58:16:58:21 | source | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:61:35:61:40 | source | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:61:35:61:40 | source | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:64:23:64:50 | composite literal | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:64:23:64:50 | composite literal | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:67:23:67:55 | call to append | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:67:23:67:55 | call to append | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:70:23:70:52 | call to append | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:70:23:70:52 | call to append | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:73:28:73:33 | source | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:73:28:73:33 | source | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:74:36:74:41 | source | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:74:36:74:41 | source | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:79:29:79:114 | ...+... | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:79:29:79:114 | ...+... | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:82:36:82:63 | call to Sprintf | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:82:36:82:63 | call to Sprintf | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:85:29:85:88 | call to Join | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:85:29:85:88 | call to Join | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:86:29:86:100 | call to Join | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:86:29:86:100 | call to Join | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:89:46:89:100 | call to Join | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:89:46:89:100 | call to Join | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
| SystemCommandExecutors.go:92:28:92:55 | call to Join | SystemCommandExecutors.go:19:12:19:18 | selection of URL : pointer type | SystemCommandExecutors.go:92:28:92:55 | call to Join | This command depends on $@. | SystemCommandExecutors.go:19:12:19:18 | selection of URL | a user-provided value |
